### ISSUE TITLE
Graphic tool to view the backward(Gradient Graph) and forward graph in Pytorch 

### ISSUE THREAD

**Issue by dinkdeep** [pytorch/pytorch#34953]
Is there any graphical tool based on dot (graphViz) similar to what (TensorFlow and Pytorch/Glow) to view the backward Graph in Pytorch or at least a way to get a textual dump of backward Graph where the Graph Tree with Nodes and there edges can be seen, somethings on the line of JIT IR. Is there some environment variable like PYTORCH_JIT_LOG_LEVEL or some logic to dump the Nodes/edges  in torch/csrc/distributed/autograd/engine/ .
-------------------
[DUMP inliner.cpp:049]   %6 : Tensor = prim::CallMethod[name="forward"](%5, %input.1)
[DUMP inliner.cpp:049]   %input.3 : Float(1, 32, 14, 14) = aten::relu(%6) # /home/ddeepani/anaconda3/lib/python3.7/site-packages/torch/nn/functional.py:914:0
[DUMP inliner.cpp:049]   %8 : Tensor = prim::CallMethod[name="forward"](%4, %input.3)
[DUMP inliner.cpp:049]   %x1 : Float(1, 64, 12, 12) = aten::relu(%8) # /home/ddeepani/anaconda3/lib/python3.7/site-packages/torch/nn/functional.py:914:0

**Comment by dinkdeep** [comment]
By mistake populated the description with comments .

**Comment by colesbury** [comment]
@szagoruyko is https://github.com/szagoruyko/pytorchviz still the best way to do this? Or is there a way with tensorboard?

**Comment by dinkdeep** [comment]
Hi Sam,  
Thanks for the link to utility, I cloned and tried the test example in the folder but did not find dot/pdf/png file which I could use to see the graph.
Some how I had to add this small function save() to make sure dot does a render operation 
and is able to write it to a pdf file.

BTW on an another side note, Is there a way we could get hold of Backward Graph in C++ code .
(https://pytorch.org/cppdocs/installing.html)

So there is a Glow backend for Pytorch (https://github.com/pytorch/glow/blob/master/torch_glow/src/Registration.cpp) which gets hold of Pytorch JIT IR for the forward graph and converts the JIT Graph Nodes to its own internal representation of the Glow nodes  and it performs its own differentiate() to generate a differential graph which kind of contains both the forward graph and backward graph (integrated together) but it does not has the way to provide access to backward graph .

Seems to me that pytorch has different forward graph and a backward graph, So essentially could there be a way to get hold of backward graph  pytorch seems to support many optimizers.

################

Hi @szagoruyko,  How did you generate the graph png listed at https://user-images.githubusercontent.com/4953728/35574234-8780297e-05d9-11e8-8e80-f4009297cefd.png"

I meant to say am I miss something obvious,  or you didn't check-in the rendering code of dot, there wasn't a Graphically view able file 

#################
import os
ROOT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
OUTPUT_DIR = os.path.join(ROOT_DIR, "test_output")
if not os.path.exists(OUTPUT_DIR):
    os.makedirs(OUTPUT_DIR)

dot = make_dot(double_backprop(x, model), params=dict(list(model.named_parameters()) + [('x', x)]))
        path = os.path.join(OUTPUT_DIR, "pytorch_dbp.pdf")
        self.save(path,dot)

    def save(self, path, dot):
        format="pdf"
        dot.format = format
        directory, file_name = os.path.split(path)
        file_name = file_name.replace("." + format, "")
        dot.render(file_name, directory=directory, cleanup=True)
########

**Comment by dinkdeep** [comment]
Hi @szagoruyko, for the forward graph with JIT am seeing some incompatibilities with  torch 1.50 python -c "import torch; print(torch.__version__)" 1.5.0a0+bc60052
It seems broken at couple of places. 
 
1) torch.onnx._optimize_trace(trace,torch._C._onnx.OperatorExportTypes.ONNX)
instead of torch._C._onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK
2) get_trace_graph is now _get_trace_graph()
trace, _ = torch.jit._get_trace_graph(model, args=(x,))
3) debugName has taken place for uniqueName 

Trying to get through them , BTW you got some bandwidth ?

**Comment by dinkdeep** [comment]
![pytroch_mlp_backward_graph](https://user-images.githubusercontent.com/61411073/77223763-6f477d80-6b85-11ea-88da-202080a376ee.png)
 @colesbury @szagoruyko ,

On a side note related to the backward graph shouldn't it look look like flipped as in the attached png, the root node is AddmmBackward (matmul )and leaf nodes are W0.bias an WO.weight and those are the ones which get populated at the end .
 
Had to inter change the edges .
   #dot.edge(str(id(u[0])), str(id(var)))
   dot.edge(str(id(var)),str(id(u[0])))

These are the nodes from add_nodes( )  and thats the way they are stream downwards while BP.
<AddmmBackward object at 0x7effa1b10090>
<AccumulateGrad object at 0x7effa1b10490>
<TanhBackward object at 0x7effa1b103d0>
<AddmmBackward object at 0x7effa2a65e10>
<AccumulateGrad object at 0x7effa1b10210>
<TBackward object at 0x7effa1b10b50>
<AccumulateGrad object at 0x7effa1b10050>
<TBackward object at 0x7effa1b10410>
<AccumulateGrad object at 0x7effa1b10450>

**Comment by linboqiao** [comment]
> ![pytroch_mlp_backward_graph](https://user-images.githubusercontent.com/61411073/77223763-6f477d80-6b85-11ea-88da-202080a376ee.png)
> @colesbury @szagoruyko ,
> 
> On a side note related to the backward graph shouldn't it look look like flipped as in the attached png, the root node is AddmmBackward (matmul )and leaf nodes are W0.bias an WO.weight and those are the ones which get populated at the end .
> 
> Had to inter change the edges .
> #dot.edge(str(id(u[0])), str(id(var)))
> dot.edge(str(id(var)),str(id(u[0])))
> 
> These are the nodes from add_nodes( ) and thats the way they are stream downwards while BP.
> <AddmmBackward object at 0x7effa1b10090>
> <AccumulateGrad object at 0x7effa1b10490>
> <TanhBackward object at 0x7effa1b103d0>
> <AddmmBackward object at 0x7effa2a65e10>
> <AccumulateGrad object at 0x7effa1b10210>
> <TBackward object at 0x7effa1b10b50>
> <AccumulateGrad object at 0x7effa1b10050>
> <TBackward object at 0x7effa1b10410>
> <AccumulateGrad object at 0x7effa1b10450>

It's pretty good, and I'm tracing ops called in C++ code too. Could you share the code to generate the backward graph? Or a text dump will also be helpful~~
